signup/signup.js

```diff
@@ -13,6 +17,198 @@ $(document).ready(function() {
+			endTime: endTime
+		});
+	}
+
+	getDate(dateTime) {
+		let t = dateTime.split(/[- :]/);
+		return new Date(t[0],t[1]-1,t[2]);
+	}
+
+	addShift(startTime, endTime) {
+		const date = this.getDate(startTime);
+		this.dates.push(date);
+		this.shiftTimes.push({
+			date: date,
+			startTime: startTime,
+			endTime: endTime
+		});
+	}
+}
+
+function loadAllSites() {
+	var request = $.ajax({
+		url: "/server/api/sites/getAll.php",
+		type: "GET",
+		dataType: "JSON",
+		data: ({
+			"siteId": true,
+			"title": true
+		}),
+		cache: false
+	})
+	.done(function(data) {
+		$siteSelect = $("#sitePicker select");
+		$siteSelect.append($('<option disabled selected value style="display:none"> -- select an option -- </option>'));
+		for(const site of data) {
+			$siteSelect.append($('<option>', {
+				value: site.siteId,
+				text : site.title
+			}));
+		}
+	});
+}
+
+// Load all of the shifts and store in a global variable
+// TODO year
+function loadAllShifts(year=(new Date()).getFullYear()) {
+	var request = $.ajax({
+		url: "/server/api/shifts/getAllShifts.php",
+		type: "GET",
+		dataType: "JSON",
+		data: ({
+			"year": year,
+			"startTime": true,
+			"endTime": true,
+			"siteId": true
+		}),
+		cache: false
+	})
+	.done(function(data) {
+		for(const shift of data) {
+			let sitesArray = sitesWithShifts.map(function(s) {return s.siteId});
+			if(sitesArray.includes(shift.siteId)) {
+				let i = sitesArray.indexOf(shift.siteId);
+				sitesWithShifts[i].addShift(shift.startTime, shift.endTime);
+			} else {
+				sitesWithShifts.push(new SiteWithShift(shift.siteId, shift.startTime, shift.endTime));
+			}
+		}
+		$("#dateInput").datepicker({
+			// Good example: https://stackoverflow.com/a/1962849/7577035
+			// called for every date before it is displayed
+			beforeShowDay: function(date) {
+				if (datesAllowed.includes(date.toDateString())) {
+					return [true, ''];
+				} else {
+					return [false, ''];
+				}
+			},
+			beforeShow: function() {
+				setTimeout(function(){
+					$('.ui-datepicker').css('z-index', 100);
+				}, 0);
+			}
+		});
+
+		// Documentation: http://jonthornton.github.io/jquery-timepicker/
+		$("#timeInput").timepicker({
+			'step': 60,
+			'forceRoundTime': true,
+			'timeFormat': 'g:i A'
+		});
+	});
+}
+
+// This function will watch the site selector and the date selector for changes.
+function updateSitesDatesAndTimes() {
+	siteSelect = $("#sitePicker select");
+	dateInput = $("#dateInput");
+	// When the value of the site selector is updated...
+	// 1. Display the date picker
+	// 2. Update the options for the date selector to disable options that are not available at this site.
+	siteSelect.change(function() {
+		$("#datePicker").show();
+		$("#timePicker").hide();
+		// Clear any previously selected date
+		$("#dateInput").val('');
+		$("#timeInput").val('');
+		var value = this.value;
+		// Clear and repopulate the array of dates that are available at the selected site
+		datesAllowed.length = 0;
```

Comment:
On a sidenote, is this a valid way to empty an array? Or does this like basically trick JS into thinking the array is empty? What is happening here?

Reply:
https://stackoverflow.com/a/1232046/7577035